aiodebug/hang_inspection.py

```diff
@@ -0,0 +1,98 @@
+from typing import List, Optional
+import sys
+import traceback
+import asyncio
+import time
+import os
+import threading
+import datetime
+
+from pygments import highlight
+from pygments.lexers import PythonLexer
+from pygments.formatters import HtmlFormatter
+
+
+
+class TraceDumper(threading.Thread):
+
+	def __init__(self, dir_name: str, interval: float, last_loop_iteration_time_wrapped: List[float], lock: threading.Lock) -> None:
+		self.interval = interval
+		self.dir_name = os.path.abspath(dir_name)
+		self.last_loop_iteration_time_wrapped = last_loop_iteration_time_wrapped
+		self.lock = lock
+		self.stop = False
+		self.monitor_task = None  # type: Optional[asyncio.Task]
+
+		# Eliminate IO errors later (on hang)
+		assert os.path.isdir(self.dir_name)
+		assert os.access(self.dir_name, os.W_OK)
+
+		threading.Thread.__init__(self)
+
+
+	def run(self) -> None:
+		'''Thread loop'''
+		while not self.stop:
+			with self.lock:
+				if time.monotonic() - self.last_loop_iteration_time_wrapped[0] > self.interval:
+					for i in range(3):
+						dt = datetime.datetime.now().strftime('%Y.%m.%d-%H:%M:%S')
+						self.save_stack_trace('stacktrace-{datetime}-{i}.html'.format(datetime = dt, i = i))
+						time.sleep(1)
+
+			time.sleep(self.interval)
+
+
+	@staticmethod
+	def _get_stack_trace() -> str:
+		code = []  # type: List[str]
+		for threadId, stack in sys._current_frames().items():
+			code.append("\n# ThreadID: %s" % threadId)
+			for filename, line_no, name, line in traceback.extract_stack(stack):
+				code.append('File: "%s", line %d, in %s' % (filename, line_no, name))
+				if line:
+					code.append("  %s" % (line.strip()))
+
+		return highlight("\n".join(code), PythonLexer(), HtmlFormatter(
+			full=False,
+			noclasses=True,
+		))
+
+
+	def save_stack_trace(self, filename: str = 'stacktrace.html') -> None:
+		with open(os.path.join(self.dir_name, filename), "wb+") as file:
```

Comment:
Use single quotes. And do we need to open the file in binary mode? If you just open it as `'w+'` you can write to it without explicitly encoding the strings below.